### ISSUE TITLE
ConsumerVersionSelectors interface does support fallbackTags (which is defined in document)

### ISSUE THREAD

**Issue by tonynguyenit18** [pact-foundation/pact-js#649]
### Software versions

- **OS**: `Mac OSX 10.15.7`
- **Consumer Pact library**:
- **Provider Pact library**: `^10.12.2`
- **Node Version**: `v12.13.0`

### Issue Checklist

Please confirm the following:

- [x] I have upgraded to the latest
- [x] I have the read the FAQs in the Readme
- [x] I have triple checked, that there are **no unhandled promises** in my code
- [x] I have set my log level to debug and attached a log file showing the complete request/response cycle
- [x] For bonus points and virtual high fives, I have created a reproduceable git repository (see below) to illustrate the problem

### Expected behaviour
Can set up like this
```js
const verificationOptions = {
  //...
  consumerVersionSelectors: [
    {
      tag: process.env.GIT_BRANCH,
      fallbackTag: "master",
      latest: true
    },
    {
      tag: "test",
      latest: true
    },
    {
      tag: "production",
      latest: true
    }
  ]
}
```

### Actual behaviour
![image](https://user-images.githubusercontent.com/37351890/117393776-0f459f00-af1f-11eb-91a3-7069d28dca2a.png)

### Steps to reproduce
I checked code base here: https://github.com/pact-foundation/pact-js-core/blob/e2d977974f7d99e027640f28bb7e69b2f0deae50/src/verifier.ts#L291
```js
export interface ConsumerVersionSelector {
  pacticipant?: string;
  tag?: string;
  version?: string;
  latest?: boolean;
  all?: boolean;
}
```
It does not match the document https://docs.pact.io/pact_broker/advanced_topics/consumer_version_selectors/

**Comment by bethesque** [comment]
We do need to fix this, but TBH, the fallback tag is unnecessary anyway, and I regret implementing it. 

This will verify main AND a matching feature branch if it exists (ensures you are backwards compatible with master)

```
{ tag: "main", latest: true }, { tag: process.env.GIT_BRANCH, latest: true }
```

This does exactly the same thing (it de-duplicates the pacts, so main pact is only verified once anyway).

```
{ tag: "main", latest: true }, { tag: process.env.GIT_BRANCH, latest: true, fallbackTag: "main" }
```

This will verify main OR a matching feature branch if it exists. This config lets you make a breaking change to master, which is actually not a good idea.

```
{ tag: process.env.GIT_BRANCH, latest: true, fallbackTag: "main" }
```

**Comment by tonynguyenit18** [comment]
Thanks @bethesque for your detailed reply.

So what the plan?
- Update the document based on current `ConsumerVersionSelectors` type? (This case we don't support fallbackTags anymore)
- OR update `ConsumerVersionSelectors` to represent document https://docs.pact.io/pact_broker/advanced_topics/consumer_version_selectors/?

**Comment by TimothyJones** [comment]
I haven’t had time to look at this yet (sorry), but I thought the JS type was mapped to a different type before it hits the wire. If that’s not the case, can we solve it with a mapping?

**Comment by tonynguyenit18** [comment]
This seem to relate to this issue https://github.com/pact-foundation/pact-js-core/issues/285

**Comment by mefellows** [comment]
> I haven’t had time to look at this yet (sorry), but I thought the JS type was mapped to a different type before it hits the wire. If that’s not the case, can we solve it with a mapping?

Yes we could, but it's not nice. I believe we do it like this:

* `"all": true` would translate to `"latest": false` (meaning you couldn't also provide `"latest": true`, we'd need to define what happens if you provide "true" to both of those options - I'd suggest we throw a validation error in this case)
* `pacticipant` translates to `consumer`
* `version` is not a thing, so we'd have to emit a deprecation warning or something

So the plan could be to implement the above mapping, and not add support for fallback tags based on Beth's advice (and future work around environments/branches support which may simplify this anyway)

**Comment by TimothyJones** [comment]
@mefellows There's a PR here that just updates the types with a breaking change https://github.com/pact-foundation/pact-js-core/pull/286

**Comment by TimothyJones** [comment]
Apologies, I misclicked. @mefellows , the PR fixes the types (thanks @tonynguyenit18 !). If we merge it in, we'll have to bump the versions for both pact-js and pact-core. This has two advantages:

1) It fixes these type issues
2) At the same time, we could deprecate pact-web, which would allow us to connect pact-js to pact-core, instead of the old package name of pact-node, simplifying the release process (https://github.com/pact-foundation/pact-js/issues/626)

It has the disadvantage that it would be "v10", which has been in alpha for a while, but won't include the v10 alpha changes. This isn't a great experience for those on the alpha release (the next release of that would then probably be `11.0.0-beta.33`).

I'm thinking it's probably best to accept the PR and do those bumps - what do you think?

**Comment by mefellows** [comment]
Thanks Tim, sounds good. I'll update the note in the other issue. I'm still not super comfortable with the new release process in the quasi-in-between state we're in. Are you OK to transition to the `11.x.x-beta` and get things back to deployable, or want to work together on it?

**Comment by bethesque** [comment]
Peeps, my original request in the master issue for the consumer version selectors was that they were *not* typed, because I want to be able to add selectors in the back end that don't require a release of the client to then support (eg. the upcoming `currentlyDeployed: true`. Is there a particular reason why they need to be typed?

**Comment by TimothyJones** [comment]
The types exist to give typescript users the IDE support and improvement to safety they expect from a typescript library. I'd prefer not to leave them untyped, because it's an implementation detail that the object is simply passed to the broker - there's an implied mapping between the pact-js DSL and the broker object (even though it's not doing any mapping)

@mefellows It turns out we can't do a major release of pact-js without including the beta content, because npm will give users of `10.0.0-beta` the `10.0.0` version when it comes out. Let's merge in the breaking change to pact-js-core, and then bump the version of pact-js-core used by pact-js v3 branch.

**Comment by bethesque** [comment]
Is there a way for a user to make use of any of the newer selectors without upgrading?

**Comment by mefellows** [comment]
Yep. I actually started responding earlier but Tim beat me to it. You can always override a Typescript type by force, and JS will let you pass anything anyway

**Comment by mefellows** [comment]
It might also be possible to specify a type that has both defined properties (i.e. the ones above) and also accept arbitrary key/value pairs, thus not making the compiler unhappy. But I prefer not, because then you don't really get good compile time safety.

**Comment by bethesque** [comment]
> Yep. I actually started responding earlier but Tim beat me to it. You can always override a Typescript type by force, and JS will let you pass anything anyway

Cool - I had that feeling, but wasn't confident enough in my knowledge of js/typescript.

**Comment by TimothyJones** [comment]
This is fixed in 10.0.0-beta.35 (apologies for the beta version - we'll get this in to master asap).
